File: packages/cli-py/src/paceproof_cli/aggregator.py
"""Verifies and aggregates records. Mirrors packages/cli-ts/src/aggregator.ts.

Security invariant (non-negotiable, see CLAUDE.md): a record with a missing,
malformed, or invalid signature is never silently included in "verified"
aggregate totals. Verified and unverified totals are computed from disjoint
lists and never merged anywhere in this module.
"""

from __future__ import annotations

from typing import cast

from .crypto import verify_record_signature
from .schema import validate_attestation_record
from .types import (
    AggregateBucket,
    AttestationRecord,
    ProviderBucket,
    RawRecord,
    ReportSummary,
    UnverifiedOutcome,
    UnverifiedReason,
    VerificationOutcome,
    VerifiedOutcome,
)


def verify_record(raw: RawRecord) -> VerificationOutcome:
    """Validates schema + Ed25519 signature for one raw record.

    Never raises -- every failure mode becomes a VerificationOutcome with a reason.
    """
    schema_result = validate_attestation_record(raw)
    if not schema_result.valid:
        return UnverifiedOutcome(
            valid=False,
            record=raw,
            reason=f"schema validation failed: {'; '.join(schema_result.errors)}",
        )

    record = cast(AttestationRecord, raw)
    sig_result = verify_record_signature(record)
    if not sig_result.valid:
        return UnverifiedOutcome(valid=False, record=raw, reason=sig_result.reason or "signature verification failed")

    return VerifiedOutcome(valid=True, record=record)


def verify_records(raws: list[RawRecord]) -> list[VerificationOutcome]:
    return [verify_record(r) for r in raws]


def _add_to_bucket_list(buckets: list[AggregateBucket], unit: str, amount: float) -> None:
    for bucket in buckets:
        if bucket["compute_unit"] == unit:
            bucket["record_count"] += 1
            bucket["compute_amount_total"] += amount
            return
    buckets.append(AggregateBucket(compute_unit=unit, record_count=1, compute_amount_total=amount))


def summarize(outcomes: list[VerificationOutcome]) -> ReportSummary:
    verified: list[VerifiedOutcome] = [o for o in outcomes if o["valid"]]
    unverified: list[UnverifiedOutcome] = [o for o in outcomes if not o["valid"]]

    verified_compute_total_by_unit: dict[str, float] = {}
    unverified_compute_total_by_unit: dict[str, float] = {}
    by_provider: dict[str, ProviderBucket] = {}
    by_workload_type: dict[str, ProviderBucket] = {}

    for verified_outcome in verified:
        r = verified_outcome["record"]
        unit = r["compute_unit"]
        verified_compute_total_by_unit[unit] = verified_compute_total_by_unit.get(unit, 0) + r["compute_amount"]

        provider_bucket = by_provider.setdefault(r["provider"], ProviderBucket(verified=[], unverified_count=0))
        _add_to_bucket_list(provider_bucket["verified"], unit, r["compute_amount"])

        workload_bucket = by_workload_type.setdefault(
            r["workload_type"], ProviderBucket(verified=[], unverified_count=0)
        )
        _add_to_bucket_list(workload_bucket["verified"], unit, r["compute_amount"])

    for unverified_outcome in unverified:
        raw = unverified_outcome["record"]
        raw_unit = raw.get("compute_unit")
        unit = raw_unit if isinstance(raw_unit, str) else "unknown_unit"
        raw_amount = raw.get("compute_amount")
        amount = raw_amount if isinstance(raw_amount, (int, float)) else 0
        unverified_compute_total_by_unit[unit] = unverified_compute_total_by_unit.get(unit, 0) + amount

        raw_provider = raw.get("provider")
        provider = raw_provider if isinstance(raw_provider, str) else "unknown_provider"
        provider_bucket = by_provider.setdefault(provider, ProviderBucket(verified=[], unverified_count=0))
        provider_bucket["unverified_count"] += 1

        raw_workload_type = raw.get("workload_type")
        workload_type = raw_workload_type if isinstance(raw_workload_type, str) else "unknown"
        workload_bucket = by_workload_type.setdefault(
            workload_type, ProviderBucket(verified=[], unverified_count=0)
        )
        workload_bucket["unverified_count"] += 1

    unverified_reasons: list[UnverifiedReason] = []
    for unverified_outcome in unverified:
        raw_record_id = unverified_outcome["record"].get("record_id")
        unverified_reasons.append(
            UnverifiedReason(
                record_id=raw_record_id if isinstance(raw_record_id, str) else None,
                reason=unverified_outcome["reason"],
            )
        )

    return ReportSummary(
        verified_count=len(verified),
        unverified_count=len(unverified),
        verified_compute_total_by_unit=verified_compute_total_by_unit,
        unverified_compute_total_by_unit=unverified_compute_total_by_unit,
        by_provider=by_provider,
        by_workload_type=by_workload_type,
        unverified_reasons=unverified_reasons,
    )

File: packages/cli-ts/src/aggregator.ts
import { validateAttestationRecord } from './schema.js';
import { verifyRecordSignature } from './crypto.js';
import type {
  AttestationRecord,
  RawRecord,
  VerificationOutcome,
  AggregateBucket,
  ReportSummary,
} from './types.js';

/**
 * Validates schema + Ed25519 signature for one raw record. A record fails
 * into "unverified" for any of: schema violation, malformed signature,
 * wrong-key signature, or tampered-payload signature. This function never
 * throws -- every failure mode becomes a VerificationOutcome with a reason.
 */
export function verifyRecord(raw: RawRecord): VerificationOutcome {
  const schemaResult = validateAttestationRecord(raw);
  if (!schemaResult.valid) {
    return { valid: false, record: raw, reason: `schema validation failed: ${schemaResult.errors.join('; ')}` };
  }

  const record = raw as unknown as AttestationRecord;
  const sigResult = verifyRecordSignature(record);
  if (!sigResult.valid) {
    return { valid: false, record: raw, reason: sigResult.reason ?? 'signature verification failed' };
  }

  return { valid: true, record };
}

export function verifyRecords(raws: RawRecord[]): VerificationOutcome[] {
  return raws.map(verifyRecord);
}

function addToBucketList(buckets: AggregateBucket[], unit: string, amount: number): void {
  const existing = buckets.find((b) => b.compute_unit === unit);
  if (existing) {
    existing.record_count += 1;
    existing.compute_amount_total += amount;
  } else {
    buckets.push({ compute_unit: unit, record_count: 1, compute_amount_total: amount });
  }
}

/**
 * Builds the report summary from verification outcomes. Verified and
 * unverified totals are computed from disjoint arrays and never merged --
 * this is the security invariant from CLAUDE.md, enforced structurally by
 * never adding an unverified record's compute_amount to any verified_*
 * accumulator anywhere in this function.
 */
export function summarize(outcomes: VerificationOutcome[]): ReportSummary {
  const verified = outcomes.filter((o): o is Extract<VerificationOutcome, { valid: true }> => o.valid);
  const unverified = outcomes.filter((o): o is Extract<VerificationOutcome, { valid: false }> => !o.valid);

  const verifiedComputeTotalByUnit: Record<string, number> = {};
  const unverifiedComputeTotalByUnit: Record<string, number> = {};
  const byProvider: ReportSummary['by_provider'] = {};
  const byWorkloadType: ReportSummary['by_workload_type'] = {};

  for (const outcome of verified) {
    const r = outcome.record;
    verifiedComputeTotalByUnit[r.compute_unit] = (verifiedComputeTotalByUnit[r.compute_unit] ?? 0) + r.compute_amount;

    byProvider[r.provider] ??= { verified: [], unverified_count: 0 };
    const providerBucket = byProvider[r.provider];
    if (providerBucket) addToBucketList(providerBucket.verified, r.compute_unit, r.compute_amount);

    byWorkloadType[r.workload_type] ??= { verified: [], unverified_count: 0 };
    const workloadBucket = byWorkloadType[r.workload_type];
    if (workloadBucket) addToBucketList(workloadBucket.verified, r.compute_unit, r.compute_amount);
  }

  for (const outcome of unverified) {
    const raw = outcome.record;
    const unit = typeof raw.compute_unit === 'string' ? raw.compute_unit : 'unknown_unit';
    const amount = typeof raw.compute_amount === 'number' ? raw.compute_amount : 0;
    unverifiedComputeTotalByUnit[unit] = (unverifiedComputeTotalByUnit[unit] ?? 0) + amount;

    const provider = typeof raw.provider === 'string' ? raw.provider : 'unknown_provider';
    byProvider[provider] ??= { verified: [], unverified_count: 0 };
    byProvider[provider].unverified_count += 1;

    const workloadType = typeof raw.workload_type === 'string' ? raw.workload_type : 'unknown';
    byWorkloadType[workloadType] ??= { verified: [], unverified_count: 0 };
    byWorkloadType[workloadType].unverified_count += 1;
  }

  return {
    verified_count: verified.length,
    unverified_count: unverified.length,
    verified_compute_total_by_unit: verifiedComputeTotalByUnit,
    unverified_compute_total_by_unit: unverifiedComputeTotalByUnit,
    by_provider: byProvider,
    by_workload_type: byWorkloadType,
    unverified_reasons: unverified.map((o) => ({
      record_id: typeof o.record.record_id === 'string' ? o.record.record_id : null,
      reason: o.reason,
    })),
  };
}
